src/bin/pubsub_server.rs

```diff
@@ -0,0 +1,170 @@
+use std::net::ToSocketAddrs;
+use std::option::Option;
+use std::sync::{Arc, RwLock};
+use std::time;
+
+#[derive(Serialize, Debug)]
+struct BlankResponse {}
+
+#[derive(Serialize, Debug)]
+struct ConnectionResponse {
+    publisher: PublisherDesc,
+    info: ConnectionInfo,
+}
+
+#[derive(Serialize, Debug)]
+struct PubSubResponse<'a, T>
+where
+    T: serde::Serialize + 'a,
+{
+    status: &'a str,
+    timestamp: time::SystemTime,
+    response: &'a T,
+}
+
+impl<'a> PubSubResponse<'a, BlankResponse> {
+    fn status_only(status: &'a str) -> PubSubResponse<'a, BlankResponse> {
+        PubSubResponse {
+            status: status,
+            timestamp: time::SystemTime::now(),
+            response: &BlankResponse {},
+        }
+    }
+}
+
+impl<'a, T> PubSubResponse<'a, T>
+where
+    T: serde::Serialize,
+{
+    fn new(status: &'a str, response: &'a T) -> PubSubResponse<'a, T> {
+        PubSubResponse {
+            status: status,
+            timestamp: time::SystemTime::now(),
+            response: response,
+        }
+    }
+}
+#[derive(Serialize, Debug)]
+struct ServiceStatus {
+    count: usize,
+}
+
+type State = HashMap<String, ConnectionResponse>;
+type ProtectedState = Arc<RwLock<State>>;
+
+fn handle_service_status_request(req: HttpRequest<ProtectedState>) -> HttpResponse {
+    let state = req.state().read().unwrap();
+    let status = ServiceStatus { count: state.len() };
+    HttpResponse::Ok().json(PubSubResponse::new("Success", &status))
+}
+
+fn handle_add_connection(
+    req: HttpRequest<ProtectedState>,
+) -> Box<Future<Item = HttpResponse, Error = Error>> {
+    let protected_state = Arc::clone(req.state());
+    req.json()
+        .from_err()
+        .and_then(move |publisher: PublisherDesc| {
+            let name = publisher.name.clone();
+            let info = ConnectionResponse {
+                publisher: publisher,
+                info: ConnectionInfo {
+                    name: name.clone(),
+                    last_report: time::SystemTime::now(),
+                },
+            };
+            let mut state = protected_state.write().unwrap();
+            state.insert(name, info);
+            Ok(HttpResponse::Ok().json(PubSubResponse::status_only("success")))
+        })
+        .responder()
+}
+
+fn handle_get_connection(info: (ActixState<ProtectedState>, Path<String>)) -> HttpResponse {
+    let (prot_state, name) = info;
+    let state = prot_state.read().unwrap();
+    match state.get(&*name) {
+        Option::None => HttpResponse::NotFound().json(PubSubResponse::status_only("not found")),
+        Option::Some(e) => HttpResponse::Ok().json(PubSubResponse::new("success", e)),
+    }
+}
+
+fn handle_get_connections(req: HttpRequest<ProtectedState>) -> HttpResponse {
+    let state = req.state().read().unwrap();
+    info!("returning connections");
+    HttpResponse::Ok().json(PubSubResponse::new("success", &*state))
+}
+
+fn actix_app() -> App<ProtectedState> {
```

Comment:
I'd rename this to make it clear it's your pubsub actix app

Reply:
In agreement.

Reply:
I missed this on the latest review. I will get it in.

Reply:
ok, fixed it.